components/region_cache_memory_engine/src/range_manager.rs

```diff
@@ -0,0 +1,260 @@
+
+impl RangeMeta {
+    fn new(id: u64) -> Self {
+        Self {
+            id,
+            range_snapshot_list: SnapshotList::default(),
+            evicted: false,
+            can_read: false,
+            safe_ts: 0,
+        }
+    }
+
+    fn derive_from(id: u64, r: &RangeMeta) -> Self {
+        Self {
+            id,
+            range_snapshot_list: SnapshotList::default(),
+            evicted: false,
+            can_read: r.can_read,
+            safe_ts: r.safe_ts,
+        }
+    }
+
+    pub(crate) fn range_snapshot_list(&self) -> &SnapshotList {
+        &self.range_snapshot_list
+    }
+}
+
+#[derive(Default)]
+struct IdAllocator(u64);
+
+impl IdAllocator {
+    fn allocate_id(&mut self) -> u64 {
+        self.0 += 1;
+        self.0
+    }
+}
+
+// RangeManger manges the ranges for RangeCacheMemoryEngine. Every new ranges
+// (whether created by new_range or by splitted due to eviction) has an unique
+// id so that range + id can exactly locate the position.
+// When an eviction occured, say we now have k1-k10 in self.ranges and the
+// eviction range is k3-k5. k1-k10 will be splitted to three ranges: k1-k3,
+// k3-k5, and k5-k10.
+// k1-k3 and k5-k10 will be new ranges inserted in self.ranges with meta dervied
+// from meta of k1-k10 (only safe_ts and can_read will be derived). k1-k10 will
+// be removed from self.ranges and inserted to self.historical_ranges. Then,
+// k3-k5 will be in the self.evicted_ranges. Now, we cannot remove the data of
+// k3-k5 as there may be some snapshot of k1-k10. After these snapshot are
+// dropped, k3-k5 can be acutally removed.
+#[derive(Default)]
+pub struct RangeManager {
+    // Each new range will increment it by one.
+    id_allocator: IdAllocator,
+    // Range before an eviction. It is recorded due to some undropped snapshot, which block the
+    // evicted range deleting the relevant data.
+    historical_ranges: BTreeMap<CacheRange, RangeMeta>,
+    evicted_ranges: BTreeSet<CacheRange>,
+    // ranges that are cached now
+    ranges: BTreeMap<CacheRange, RangeMeta>,
+}
+
+impl RangeManager {
+    pub(crate) fn ranges(&self) -> &BTreeMap<CacheRange, RangeMeta> {
+        &self.ranges
+    }
+
+    pub(crate) fn new_range(&mut self, range: CacheRange) {
+        let range_meta = RangeMeta::new(self.id_allocator.allocate_id());
+        self.ranges.insert(range, range_meta);
+    }
+
+    pub fn set_range_readable(&mut self, range: &CacheRange, set_readable: bool) {
+        if let Some(meta) = self.ranges.get_mut(range) {
+            meta.can_read = true;
+        }
+    }
+
+    pub fn set_safe_ts(&mut self, range: &CacheRange, safe_ts: u64) -> bool {
+        if let Some(meta) = self.ranges.get_mut(range) {
+            if meta.safe_ts > safe_ts {
+                return false;
+            }
+            meta.safe_ts = safe_ts;
+            true
+        } else {
+            false
+        }
+    }
+
+    pub(crate) fn overlap_with_range(&self, range: &CacheRange) -> bool {
```

Comment:
Can we also add `contains(&self, key: &[u8])` or `get_range_meta_for_key(self, key; &[u8]) -> Option<RangeMeta>`?

Reply:
what's the use cases?

Reply:
Ok, get it.